mdal/frmts/mdal_esri_tin.cpp

```diff
@@ -0,0 +1,300 @@
+    if ( !readValue( vert.y, inXY, isNativeLittleEndian ) )
+      throw MDAL_Status::Err_UnknownFormat;
+
+    float zValue;
+    if ( !readValue( zValue, inZ, isNativeLittleEndian ) )
+      throw MDAL_Status::Err_UnknownFormat;
+    vert.z = double( zValue );
+
+    if ( correctedIndex( fileIndex, superpointIndexes ) >= 0 ) //store the vertex only if it is not a superpoint
+      vertices.push_back( vert );
+    fileIndex++;
+  }
+
+  inXY.close();
+  inZ.close();
+}
+
+void MDAL::DriverEsriTin::populateFaces( const std::string &uri, MDAL::Faces &faces, const std::list<int> &superpointIndexes ) const
+{
+  bool isNativeLittleEndian = MDAL::isNativeLittleEndian();
+  std::ifstream inFaces( faceFile( uri ), std::ifstream::in | std::ifstream::binary );
+  std::ifstream inMsk( mskFile( uri ), std::ifstream::in | std::ifstream::binary );
+  std::ifstream inMsx( msxFile( uri ), std::ifstream::in | std::ifstream::binary );
+
+  if ( ! inFaces.is_open() )
+    throw MDAL_Status::Err_FileNotFound;
+  if ( ! inMsk.is_open() )
+    throw MDAL_Status::Err_FileNotFound;
+  if ( ! inMsx.is_open() )
+    throw MDAL_Status::Err_FileNotFound;
+
+  //Find the beginning of data in mskFile
+  inMsx.seekg( -4, std::ios::end );
+  int32_t mskBegin;
+  if ( ! readValue( mskBegin, inMsx, true ) )
+    throw  MDAL_Status::Err_UnknownFormat;
+
+  //read information in mskFile
+  inMsk.seekg( -mskBegin * 2, std::ios::end );
+  int32_t maskIntergerCount;
+  if ( ! readValue( maskIntergerCount, inMsk, true ) )
+    throw MDAL_Status::Err_UnknownFormat;
+  inMsk.ignore( 4 ); //unused 4 bytes
+  int32_t maskBitsCount;
+  if ( ! readValue( maskBitsCount, inMsk, true ) )
+    throw MDAL_Status::Err_UnknownFormat;
+
+  int c = 0;
+  int32_t maskInt = 0;
+  while ( true )
+  {
+    //read mask file
+    if ( c % 32 == 0 && c < maskBitsCount ) //first bit in the mask array have to be used-->read next maskInt
+      if ( ! readValue( maskInt, inMsk, true ) )
+        throw MDAL_Status::Err_UnknownFormat;
+
+    Face f;
+    for ( int i = 0; i < 3; ++i )
+    {
+      int32_t index;
+      if ( ! readValue( index, inFaces, isNativeLittleEndian ) )
+        break;
+
+      index = correctedIndex( index, superpointIndexes );
+
+      f.push_back( static_cast<size_t>( index ) );
+    }
+
+    if ( f.size() == 0 ) //that's mean this is the end of the file
+      break;
+
+    if ( f.size() < 3 ) //that's mean the face is not complete
+      throw MDAL_Status::Err_UnknownFormat;
+
+    //exclude masked face
+    if ( !( maskInt & 0x01 ) )
+      faces.push_back( f );
+
+    c++;
+    maskInt = maskInt >> 1;
+  }
+
+  inFaces.close();
+  inMsk.close();
+  inMsx.close();
+}
+
+std::string MDAL::DriverEsriTin::getCrsWkt( const std::string &uri ) const
+{
+  std::ifstream inCRS( crsFile( uri ), std::ifstream::in );
+  if ( ! inCRS.is_open() )
+    return std::string();
+
+  std::string crsWkt;
+  std::getline( inCRS, crsWkt );
+
+  if ( crsWkt == "{B286C06B-0879-11D2-AACA-00C04FA33C20}" )
```

Comment:
I do not understand this {B28....}? 

Wouldn't it be more sensible to read projection from *.prj file? (3D_Boreholes.prj in the test case?)

Reply:
It's a "magic sequence" used in ESRI to indicate a "not set" or unknown CRS. (It's actually the com class id of the esri UnknownCoordinateSystem class).

Reply:
do you know if this file contains actually WKT or it is always some magic sequence of class ids?

Reply:
I don't know why the *.prj file is here. With other TIN, there is not this file but there is always a prj.adf file. The prj.adf file contains WKT.

Reply:
ok no problem
, can you add comment on the line?

Reply:
Done